Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+            let sequenceNumber : UInt32 = UInt32((Int(bitPattern) >> 32) & 0x00FFFFFF)
+            let eventType = CqeEventType(rawValue:((Int(bitPattern) >> 32) & 0xFF000000) >> 24) // shift out the fd
+
+            let bitpatternAsPointer = UnsafeMutableRawPointer.init(bitPattern: bitPattern)
+
+            _debugPrintCQE("\(i) = fd[\(fd)] eventType[\(String(describing:eventType))] sequenceNumber[\(sequenceNumber)] res [\(c.res)] flags [\(c.flags)]  bitpattern[\(String(describing:bitpatternAsPointer))]")
+        }
+        #endif
+    }
+
+    init() {
+        cqes = UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>.allocate(capacity: Int(cqeMaxCount))
+        cqes.initialize(repeating:&emptyCqe, count:Int(cqeMaxCount))
+    }
+
+    deinit {
+        cqes.deallocate()
+    }
+
+    internal func io_uring_queue_init() throws -> () {
+        // IORING_SETUP_SQPOLL will be fundamentally useful in 5.13, as we no longer need elevated privileges as previously
+        // and CNIOLinux.io_uring_queue_init will setup a shared uring instance which allows us to use a single
+        // kernel sqpoll thread that is shared amongst all loops. NB, each process will have it's own polling thread
+        // if SQPOLL is enabled, so for hosts running multiple processes using it, some care should be taken to not overload.
+        #if SWIFTNIO_LIBURING_SQPOLL_ENABLED
+        let _sqpollEnabled : UInt32 = 1
+        #else
+        let _sqpollEnabled : UInt32 = 0
+        #endif
+        if (CNIOLinux_io_uring_queue_init(ringEntries, &ring, _sqpollEnabled ) != 0)
+         {
+             throw UringError.uringSetupFailure
+         }
+
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+
+    internal func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux.io_uring_queue_exit(&ring)
+    }
+
+    static internal func io_uring_use_multishot_poll() -> Bool {
+        #if SWIFTNIO_LIBURING_MULTISHOT
+        return true
+        #else
+        return false
+        #endif
+    }
+
+    // Adopting some retry code from queue.c from liburing with slight
+    // modifications - we never want to have to handle retries of
+    // SQE allocation in all places it could possibly occur.
+    // If the SQ ring is full, we may need to submit IO first
+    internal func get_sqe() -> UnsafeMutablePointer<io_uring_sqe>?
+    {
+        // io_uring_submit can fail here due to backpressure from kernel for not reaping CQE:s.
+        //
+        // I think we should consider handling that as a fatalError, as fundamentally the ring size is too small
+        // compared to the amount of events the user tries to push through in a single eventloop tick.
+        //
+        // This is mostly a problem for synthetic tests that e.g. do a huge amount of registration modifications.
+        //
+        // This is a slight design issue with SwiftNIO in general that should be discussed.
```

Comment:
Sounds good. Let's discuss this.

Reply:
I think (as a separate issue) it might be worth considering introducing some back pressure mechanism for registration modifications (that really only would kick in for degenerate cases), or alternatively (what I probably would prefer) is that we'd start tracking the registration modifications and merge/coalesce all of them that are done during a single event loop tick. Then we'd have an upper bound of `numfds` (channels) of outbound registration change events which would not fill up e.g. the outbound SQE queue and make less system calls for such cases.  Then again, it is mostly an issue for tests, but it would be nice to handle degenerate behaviours too robustly. I don't know what you think? Should probably move on to a separate issue anyway.

Reply:
@weissi @Lukasa - opened https://github.com/apple/swift-nio/issues/1832 for that, hope it is clear enough.